frontend/lib/resolution/scope-queries.cpp

```diff
@@ -1530,6 +1560,29 @@ static void errorIfNameNotInScope(Context* context,
     return;
   }
 
+  if ((foundViaPrivate || foundViaBuiltin) &&
+      parsing::shouldWarnUnstableForId(context, exprForError->id())) {
+    // try to avoid throwing the warning if the user imports a (private) method
+    // from a module. If that's what they're doing, it's not really a tertiary
+    // import, it's just an error, which they will be shown later.
+    bool shouldWarnForTertiaryImport = false;
+    if (foundViaBuiltin) {
+      shouldWarnForTertiaryImport = true;
+    } else if (foundViaPrivate) {
+      for (auto& ids : result) {
+        for (auto& id : ids) {
+          if (asttags::isTypeDecl(parsing::idToTag(context, id))) {
```

Comment:
Could you explain this logic to me? I'm reading the comment as saying we don't want to warn when the user directly imports a private method by name. And I see this code as only warning if the use/import causes us to bring in a private type declaration. But wouldn't that exclude private tertiary methods?

Reply:
The dyno scope resolver doesn't actually bring in teriary methods in this way right now; that task is left up to the production scope resolver. In fact, I implemented exactly the following sentence from the email:

> mark `import M.t` and `use M only t` unstable if `t` is not defined in `M` 

So if you're importing a type from a module, but that type isn't defined in the module, the only thing you can be doing is importing the tertiary methods on that type. Hence, this warning checks for "importing a type not defined in M" and warns then.

Reply:
I see, ok. Does that mean we warn even if no tertiary methods are actually brought in?

Reply:
Yes, but I think that's reasonable: the syntactic form `import M.sometype` if `sometype` is not in `M` is used expressly for the purpose of tertiary method imports. I suppose I could add a note to the error message to explain why it's being flagged as a tertiary import, in case the user was simply trying to get ahold of `sometype`.